vault/keysource.go

```diff
@@ -0,0 +1,241 @@
+func (key *MasterKey) SetEncryptedDataKey(enc []byte) {
+	key.EncryptedKey = string(enc)
+}
+
+// Encrypt takes a sops data key, encrypts it with Vault Transit and stores the result in the EncryptedKey field
+func (key *MasterKey) Encrypt(dataKey []byte) error {
+	path := filepath.Join(key.BackendPath, "encrypt", key.KeyName)
+	cfg := api.DefaultConfig()
+	cfg.Address = key.VaultAddress
+	cli, err := api.NewClient(cfg)
+	if err != nil {
+		log.WithField("Path", path).Info("Vault connections failed")
+		return fmt.Errorf("Cannot create Vault Client: %v", err)
+	}
+	encoded := base64.StdEncoding.EncodeToString(dataKey)
+	payload := make(map[string]interface{})
+	payload["plaintext"] = encoded
+	raw, err := cli.Logical().Write(path, payload)
+	if err != nil {
+		log.WithField("Path", path).Info("Encryption failed")
+		return err
+	}
+	if raw == nil || raw.Data == nil {
+		return fmt.Errorf("The transit backend %s is empty", path)
+	}
+	encrypted, ok := raw.Data["ciphertext"]
+	if ok != true {
+		return fmt.Errorf("there's not encrypted data")
+	}
+	encryptedKey, ok := encrypted.(string)
+	if ok != true {
+		return fmt.Errorf("the ciphertext cannot be casted to string")
+	}
+	key.EncryptedKey = encryptedKey
+	return nil
+}
+
+// EncryptIfNeeded encrypts the provided sops' data key and encrypts it if it hasn't been encrypted yet
+func (key *MasterKey) EncryptIfNeeded(dataKey []byte) error {
+	if key.EncryptedKey == "" {
+		return key.Encrypt(dataKey)
+	}
+	return nil
+}
+
+// Decrypt decrypts the EncryptedKey field with CGP KMS and returns the result.
+func (key *MasterKey) Decrypt() ([]byte, error) {
+	path := filepath.Join(key.BackendPath, "decrypt", key.KeyName)
+	cfg := api.DefaultConfig()
+	cfg.Address = key.VaultAddress
+	cli, err := api.NewClient(cfg)
+	if err != nil {
+		log.WithField("Path", path).Info("Vault connections failed")
+		return nil, fmt.Errorf("Cannot create Vault Client: %v", err)
+	}
+	payload := make(map[string]interface{})
+	payload["ciphertext"] = key.EncryptedKey
+	raw, err := cli.Logical().Write(path, payload)
+	if err != nil {
+		log.WithField("Path", path).Info("Encryption failed")
+		return nil, err
+	}
+	if raw == nil || raw.Data == nil {
+		return nil, fmt.Errorf("The transit backend %s is empty", path)
+	}
+	decrypted, ok := raw.Data["plaintext"]
+	if ok != true {
+		return nil, fmt.Errorf("there's no decrypted data")
+	}
+	dataKey, ok := decrypted.(string)
+	if ok != true {
+		return nil, fmt.Errorf("the plaintest cannot be casted to string")
+	}
+	result, err := base64.StdEncoding.DecodeString(dataKey)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't decode base64 plaintext")
+	}
+	return result, nil
+}
+
+// NeedsRotation returns whether the data key needs to be rotated or not.
+// This is simply copied from GCPKMS
+// TODO: handle key rotation on vault side
+func (key *MasterKey) NeedsRotation() bool {
+	//TODO: manage rewrapping https://www.vaultproject.io/api/secret/transit/index.html#rewrap-data
```

Comment:
That's interesting! I think it'd require changes to the `MasterKey` interface to make it useful, but it's good to know that there's encryption backends that allow this.